compiler/rustc_const_eval/src/transform/promote_consts.rs

```diff
@@ -272,40 +279,62 @@ impl<'tcx> Validator<'_, 'tcx> {
-                            statement.source_info.span,
-                            "{:?} is not an assignment",
-                            statement
-                        );
-                    }
-                }
-            } else {
-                let terminator = block.terminator();
-                match &terminator.kind {
-                    TerminatorKind::Call { func, args, .. } => self.validate_call(func, args),
-                    TerminatorKind::Yield { .. } => Err(Unpromotable),
-                    kind => {
-                        span_bug!(terminator.source_info.span, "{:?} not promotable", kind);
-                    }
+    fn validate_local(&mut self, local: Local) -> Result<(), Unpromotable> {
+        if let TempState::Defined { location: loc, uses, valid } = self.temps[local] {
+            match valid {
+                Validity::Invalid => Err(Unpromotable),
+                Validity::Validated => Ok(()),
+                Validity::Unknown => {
+                    let ok = {
+                        let block = &self.body[loc.block];
+                        let num_stmts = block.statements.len();
+
+                        if loc.statement_index < num_stmts {
+                            let statement = &block.statements[loc.statement_index];
+                            match &statement.kind {
+                                StatementKind::Assign(box (_, rhs)) => self.validate_rvalue(rhs),
+                                _ => {
+                                    span_bug!(
+                                        statement.source_info.span,
+                                        "{:?} is not an assignment",
+                                        statement
+                                    );
+                                }
+                            }
+                        } else {
+                            let terminator = block.terminator();
+                            match &terminator.kind {
+                                TerminatorKind::Call { func, args, .. } => {
+                                    self.validate_call(func, args)
+                                }
+                                TerminatorKind::Yield { .. } => Err(Unpromotable),
+                                kind => {
+                                    span_bug!(
+                                        terminator.source_info.span,
+                                        "{:?} not promotable",
+                                        kind
+                                    );
+                                }
+                            }
+                        }
+                    };
+                    self.temps[local] = TempState::Defined {
+                        location: loc,
+                        uses,
+                        valid: match ok {
+                            Ok(()) => Validity::Validated,
+                            Err(_) => Validity::Invalid,
```

Comment:
Considering the `Validity` enum is just being converted back and forth between Result and itself, maybe we should just use `Option<Result<(), Unpromotable>>` instead. I usually like specialized enums, but in this case I think it's not very helpful.

Reply:
Hmm... actually, in case of `Err(_)`, I think we may just want to replace `self.temps[local]` with `TempState::Unpromotable`.

Reply:
So... basically we don't need the `Invalid` variant, because we can just change the entire `TempState`.

That leaves us with `Validated` and `Unknown`.

Now I'm just worried that `validate_rvalue` and `validate_call` may switch from an `Ok` to an `Unpromotable` if run again. Idea: if `cfg!(debug_assertions)` is on, always run the `Unknown` path, but then, if we already had a cached `Validated`, check that we would have set it to `Validated` again and `span_bug!` otherwise.

Reply:
I don't think we need to worry about this. 
1. `validate_local` only checks the statement which an `local` first appears (and as an lvalue), so the `local` will not appear in the rvalue or the function parameters of this statement. 
2. During the validate_* process, `validate_local` is always executed when `local` is encountered, so their results are always the same.

Reply:
> Hmm... actually, in case of `Err(_)`, I think we may just want to replace `self.temps[local]` with `TempState::Unpromotable`.

That makes sense!